File: python/010.py
#   010
#   Regular Expression Matching
#   2015-04-18
#####################################################
class Solution:
    # @param s, a string
    # @param p, a string
    # @return a boolean
    def isMatch(self, s, p):
        l1 = len(s)
        l2 = len(p)

        i = 0
        j = 0
        while i < l1 and j < l2:
            if j+1 < l2 and p[j+1] == '*':
                if s[i] != p[j] and p[j] != '.':
                    j+=2
                elif self.isMatch(s[i:], p[j+2:]):
                    return True
                elif s[i] == p[j] or p[j] == '.':
                    i+=1
                else:
                    return False
            elif s[i] == p[j] or p[j] == '.':
                i+=1
                j+=1
            else:
                return False
        if i < l1:
            return False
        while j < l2 :
            if j+1 < l2 and p[j+1] == '*':
                j+=2
        return j == l2

#####################################################
    # @param s, a string
    # @param p, a string
    # @return a boolean
    def isMatch2(self, s, p):
        l1 = len(s)
        l2 = len(p)
        m = [True] + [False] * l1
        for j in range(0, l2):
            if j+1 < l2 and p[j+1] == '*':
                continue
            elif p[j] == '*':
                c = p[j-1]
                for i in range(1, l1+1):
                    m[i] = m[i] or (m[i-1] and (s[i-1] == c or c == '.'))
            else:
                c = p[j]
                for i in range(l1, 0, -1):
                    m[i] = m[i-1] and (m[i-1] and (s[i-1] == c or c == '.'))
                m[0] = False
        return m[l1]

#####################################################

s = "aaa";
p = "a*a";
s = "ab";
p = ".*..c*";
s = "baccbbcbcacacbbc";
p = "c*.*b*c*ba*b*b*.a*";
s = "aaa"
p = "aaaa"

print(s)
print(p)

sol = Solution()
z = sol.isMatch2(s, p)
print(z)

File: c/010.c
//
//  010.c
//  leetcode
//
//  Created by Hang Zhang on 4/17/15.
//  Copyright (c) 2015 Hilings Studio. All rights reserved.
//

#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

bool isMatch(char* s, char* p) {
    while (*p) {
        if (*(p+1) == '*') {
            if (*s != *p && *p != '.') {
                p+=2;
            } else if (isMatch(s, p+2)) {
                return true;
            } else if (*s == *p || (*p == '.' && *s != '\0')) {
                s++;
            } else {
                return false;
            }
        } else if (*s == *p || (*p == '.' && *s != '\0')) {
            s++;
            p++;
        } else {
            return false;
        }
    }
    return *s == '\0' && *p == '\0';
}

bool isMatch2(char *s, char *p){
    int l1 = (int)strlen(s);
    bool* m = malloc(sizeof(bool) * (l1+1));
    m[0] = true;
    for (int i = 1; i <= l1; i++) {
        m[i] = false;
    }

    for (; *p != '\0'; p++) {
        if (*(p+1) == '*') {
            continue;
        } else if (*p == '*') {
            char c = *(p-1);
            for (int i = 1; i <= l1; i++) {
                m[i] = m[i] || (m[i - 1] && (s[i - 1] == c || c == '.'));
            }
        } else {
            char c = *p;
            for (int i = l1; i > 0; i--) {
                m[i] = m[i - 1] && (s[i - 1] == c || c == '.');
            }
            m[0] = false;
        }
    }

    bool r = m[l1];
    free(m);
    return r;
}


int main(int arg, char *argv[]) {
    // insert code here...
    printf("LeetCode 010 Regular Expression Matching, C ... ...\n");
    /*
     isMatch("aa","a") → false
     isMatch("aa","aa") → true
     isMatch("aaa","aa") → false
     isMatch("aa", "a*") → true
     isMatch("aa", ".*") → true
     isMatch("ab", ".*") → true
     isMatch("aab", "c*a*b") → true
     */
    char s[] = "";
    char p[] = "a";
    //char s[] = "aaa";
    //char p[] = "a*a";
    //char s[] = "ab";
    //char p[] = ".*..c*";
    //char s[] = "baccbbcbcacacbbc";
    //char p[] = "c*.*b*c*ba*b*b*.a*";
    //char s[] = "aa";
    //char p[] = ".*";

    printf("%s  %s\n", s, p);
    printf("%s\n", isMatch2(s, p) ? "true" : "false");
    
    return 0;
}